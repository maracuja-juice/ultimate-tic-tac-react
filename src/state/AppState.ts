import { Point } from '../util';

export interface AppState {
  currentGame: GameInformation;
  finishedGames: FinishedGameState[];
  analysisGame: AnalysisGame;
}

export interface GameInformation {
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
}

export interface FinishedGameState {
  id?: string;
  date: string;
  winner: Winner;
  gameState: SmallBoardInformation[];
  moves: MoveState[];
  saveState: string;
  errorMessage: string;
}

export interface AnalysisGame {
  id: string;
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
  currentMove: number;
}

export interface GameState {
  currentPlayer: Player;
  isFinished: boolean;
  winningPlayer: Winner;
}

export interface TileInformation {
  position: Point;
  value: TileValue;
}

export interface SmallBoardInformation extends TileInformation {
  tiles: SmallTileInformation[];
}

export interface SmallTileInformation extends TileInformation {
  boardPosition: Point;
}

export enum TileValue {
  Cross = 0,
  Circle = 1,
  Empty = 2,
  Destroyed = 3,
}

export enum Winner {
  Cross = 0,
  Circle = 1,
  None = 2,
  Draw = 3,
}

export enum Player {
  Cross = 0,
  Circle = 1,
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}

export interface MoveState extends Move {
  player: Player;
  moveNumber: number;
}

export interface GenericAction {
  type: string;
  // tslint:disable-next-line: no-any
  payload?: any;
}

export interface MarkSpecially {
  condition: boolean;
  position?: { boardPosition: Point; tilePosition: Point };
}
