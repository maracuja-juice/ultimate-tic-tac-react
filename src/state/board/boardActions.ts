import { TileValue } from '../AppState';
import { Point } from '../../util/Point';

export const SET_TILE_VALUE = 'tile-reducer/set-tile-value';
export const SET_BOARD_VALUE = 'tile-reducer/set-board-value';
export const CALCULATE_BOARD_VALUE = 'tile-reducer/calculate-board-value';

export const setTileValue = ( boardPosition: Point, tilePosition: Point, tileValue: TileValue ) => ({
    type: SET_TILE_VALUE,
    payload: {boardPosition: boardPosition, tilePosition: tilePosition, tileValue: tileValue}
});

export const setBoardValue = ( boardPosition: Point, tileValue: TileValue ) => ({
    type: SET_BOARD_VALUE,
    payload: {boardPosition: boardPosition, tileValue}
});

export const calculateBoardValue = ( boardPosition: Point ) => ({
    type: CALCULATE_BOARD_VALUE,
    payload: boardPosition
});