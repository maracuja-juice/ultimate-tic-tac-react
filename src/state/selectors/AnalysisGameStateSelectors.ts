import { AnalysisGame, AppState, Player } from '../AppState';

export const getAnalysisGame = (
  state: AppState,
  id: string,
): AnalysisGame | undefined => {
  const finishedGame = state.finishedGames.find((g) => g.id === id);
  if (finishedGame) {
    return {
      id: finishedGame.id!,
      board: finishedGame.gameState,
      moves: finishedGame.moves,
      currentMove: finishedGame.moves.length,
      game: {
        currentPlayer: Player.Circle,
        winningPlayer: finishedGame.winner,
        isFinished: true,
      },
      activeBoards: [],
    };
  }

  return undefined;
};
