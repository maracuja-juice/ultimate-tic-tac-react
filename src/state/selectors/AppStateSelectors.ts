import { AnalysisGame, AppState, Player } from '../AppState';

export const getBoards = (state: AppState) => state.currentGame.board;

export const getCurrentPlayer = (state: AppState) =>
  state.currentGame.game.currentPlayer;

export const getWinningPlayer = (state: AppState) =>
  state.currentGame.game.winningPlayer;

export const getMoves = (state: AppState) => state.currentGame.moves;

export const getFinishedGame = (
  state: AppState,
  id: string,
): AnalysisGame | undefined => {
  const finishedGame = state.finishedGames.find((g) => g.id === id);
  if (finishedGame) {
    return {
      moves: finishedGame.moves,
      currentMove: finishedGame.moves.length,
      game: {
        currentPlayer: Player.Circle,
        winningPlayer: finishedGame.winner,
        isFinished: true,
      },
      activeBoards: [],
      board: finishedGame.gameState,
      id: finishedGame.id!,
    };
  }

  return undefined;
};
