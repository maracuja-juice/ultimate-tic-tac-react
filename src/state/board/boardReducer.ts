import { cloneState, GenericAction, SmallBoardInformation, SmallTileInformation, TileValue } from '../AppState';
import { SET_BOARD_VALUE, SET_TILE_VALUE } from './boardActions';
import { arePointsEqual, Point } from '../../util/Point';
import { RESTART_GAME } from '../commonAction';

const getSmallBoardTiles = ( boardX: number, boardY: number ) => {
    let tiles: SmallTileInformation[] = [];
    for (let x = 0; x < 3; x++) {
        for (let y = 0; y < 3; y++) {
            tiles.push( {
                            boardPosition: {x: boardX, y: boardY},
                            position: {x, y},
                            value: TileValue.Empty
                        } );
        }
    }
    return tiles;
};

const getInitialState = () => {
    let state: SmallBoardInformation[] = [];
    for (let x = 0; x < 3; x++) {
        for (let y = 0; y < 3; y++) {
            state.push( {
                            value: TileValue.Empty,
                            position: {x, y},
                            tiles: getSmallBoardTiles( x, y )
                        } );
        }
    }
    return state;
};

const getSmallBoard = (bigBoard: SmallBoardInformation[], boardPosition: Point) => {
    const smallBoardIndex = bigBoard.findIndex( board => {
        return arePointsEqual( board.position, boardPosition );
    } );
    return bigBoard[smallBoardIndex];
};

const initialState = getInitialState();

const boardReducer = ( state = initialState, action: GenericAction ) => {
    switch (action.type) {

        case SET_TILE_VALUE: {
            let clone = cloneState( state );

            const smallBoard = getSmallBoard(clone, action.payload.boardPosition);
            const tileIndex = smallBoard.tiles.findIndex( tile => {
                return arePointsEqual( tile.position, action.payload.tilePosition );
            } );
            smallBoard.tiles[tileIndex].value = action.payload.tileValue;

            return clone;
        }
        case SET_BOARD_VALUE: {
            let clone = cloneState( state );

            const smallBoard = getSmallBoard(clone, action.payload.boardPosition);
            smallBoard.value = action.payload.tileValue;

            return clone;
        }
        case RESTART_GAME: {
            return initialState;
        }
    }

    return state;
};

export default boardReducer;
