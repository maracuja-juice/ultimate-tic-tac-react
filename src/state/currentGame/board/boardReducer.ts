import { GenericAction, SmallBoardInformation } from '../../AppState';
import produce from 'immer';
import { SET_BOARD_VALUE, SET_TILE_VALUE } from './boardActions';
import { RESTART_GAME } from '../../commonAction';
import { TicTacToeGame, Point, arePointsEqual } from '../../../util';

const getSmallBoard = (
  bigBoard: SmallBoardInformation[],
  boardPosition: Point,
) => {
  const smallBoardIndex = bigBoard.findIndex((board) => {
    return arePointsEqual(board.position, boardPosition);
  });
  return bigBoard[smallBoardIndex];
};

const initialState = new TicTacToeGame([]).getBoard();

const boardReducer = (state = initialState, action: GenericAction) => {
  switch (action.type) {
    case SET_TILE_VALUE: {
      const newBoardState = produce(state, (draftState) => {
        const smallBoard = getSmallBoard(
          draftState,
          action.payload.boardPosition,
        );
        const tileIndex = smallBoard.tiles.findIndex((tile) => {
          return arePointsEqual(tile.position, action.payload.tilePosition);
        });
        smallBoard.tiles[tileIndex].value = action.payload.tileValue;
      });

      return newBoardState;
    }
    case SET_BOARD_VALUE: {
      const newBoardState = produce(state, (draftState) => {
        const smallBoard = getSmallBoard(
          draftState,
          action.payload.boardPosition,
        );
        smallBoard.value = action.payload.tileValue;
      });

      return newBoardState;
    }
    case RESTART_GAME: {
      return initialState;
    }

    default: {
      return state;
    }
  }
};

export default boardReducer;
