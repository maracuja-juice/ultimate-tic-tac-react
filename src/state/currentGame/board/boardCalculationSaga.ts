import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction } from '../../AppState';
import { CALCULATE_BOARD_VALUE, setBoardValue } from './boardActions';
import { playerToTileValue, TicTacToeGame } from '../../../lib';
import { getMoves } from '../../selectors/appStateSelectors';
import { SagaIterator } from 'redux-saga';

function* calculateWinningBoard(action: GenericAction): SagaIterator {
  const boardPosition = action.payload;

  const moves = yield select(getMoves);

  const game = new TicTacToeGame(moves);
  const winResult = game.getWinResultForSmallBoard(boardPosition);
  if (winResult.isFinished) {
    const newSmallBoardTileValue = playerToTileValue(
      winResult.winningPlayer,
      true,
    );
    yield put(setBoardValue(boardPosition, newSmallBoardTileValue));
  }
}

function* boardCalculationSaga() {
  yield takeEvery(CALCULATE_BOARD_VALUE, calculateWinningBoard);
}

export default boardCalculationSaga;
