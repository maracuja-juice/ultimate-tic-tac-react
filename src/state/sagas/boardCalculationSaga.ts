import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction, SmallBoardInformation } from '../AppState';
import { CALCULATE_BOARD_VALUE, setBoardValue } from '../board/boardActions';
import { getWinResult } from '../../util/CheckBoard';
import { arePointsEqual, playerToTileValue } from '../../util';
import { getBoards } from '../selectors/AppStateSelectors';

function* calculateWinningBoard( action: GenericAction ) {
    const boardPosition = action.payload;

    const boards = yield select( getBoards );

    const affectedBoard = boards.find(
        ( board: SmallBoardInformation ) =>
            arePointsEqual( board.position, boardPosition )
    ).tiles;

    const winResult = getWinResult( affectedBoard );
    if (winResult.isFinished) {
        const newSmallBoardTileValue = playerToTileValue( winResult.winningPlayer!, true );
        yield put( setBoardValue( boardPosition, newSmallBoardTileValue ) );
    }
}

function* boardCalculationSaga() {
    yield takeEvery( CALCULATE_BOARD_VALUE, calculateWinningBoard );
}

export default boardCalculationSaga;