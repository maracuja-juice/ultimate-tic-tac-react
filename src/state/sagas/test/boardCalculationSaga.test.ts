import { expectSaga } from 'redux-saga-test-plan';
import boardCalculationSaga from '../boardCalculationSaga';
import { getBoards } from '../../selectors/AppStateSelectors';
import { select } from 'redux-saga/effects';
import { CALCULATE_BOARD_VALUE, SET_BOARD_VALUE } from '../../board/boardActions';
import boardStateMock from './boardMock.test';
import { Player } from '../../AppState';

describe( 'boardCalculationSaga', () => {
    it( 'should dispatch at least the following actions, order not tested', () => {
        return expectSaga( boardCalculationSaga )
            .provide([
                         [select(getBoards), boardStateMock]
                     ])
            .put({type: SET_BOARD_VALUE, payload: {boardPosition: {x: 1, y: 0}, tileValue: Player.Cross}})
            .dispatch( {type: CALCULATE_BOARD_VALUE, payload: {x: 1, y: 0}} ) // TODO maybe this needs to be switched to 0,1
            .run();
    } );

    // if more put effects happen: this catches it + this checks for the order
    it('should match snapshot', () => {
        return expectSaga(boardCalculationSaga)
            .provide([
                         [select(getBoards), boardStateMock]
                     ])
            .dispatch( {type: CALCULATE_BOARD_VALUE, payload: {x: 1, y: 0}} )
            .run()
            .then((result) => {
                expect(result.toJSON()).toMatchSnapshot();
            });
    });
} );
