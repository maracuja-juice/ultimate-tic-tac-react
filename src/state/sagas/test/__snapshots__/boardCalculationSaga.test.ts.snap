// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`boardCalculationSaga should match snapshot 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "tile-reducer/calculate-board-value",
          "@@redux-saga-test-plan/json/function/calculateWinningBoard",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          Object {
            "payload": Object {
              "x": 1,
              "y": 0,
            },
            "type": "tile-reducer/calculate-board-value",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {
            "boardPosition": Object {
              "x": 1,
              "y": 0,
            },
            "tileValue": 0,
          },
          "type": "tile-reducer/set-board-value",
        },
        "channel": null,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "tile-reducer/calculate-board-value",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "tile-reducer/calculate-board-value",
      },
    },
  ],
}
`;
